src/main/rascal/lang/jimple/toolkit/ssa/VariableRenaming.rsc

```diff
@@ -0,0 +1,499 @@
+public FlowGraph applyVariableRenaming(FlowGraph flowGraph) {
+	VARIABLE_VERSION_STACK = (); // Stack for each variable that holds the next variable to ve replaced
+	VARIABLE_ASSIGNMENT_COUNT = (); // Counts how many assingments have been processed for a given variable
+	REPLACED_NODES = {}; // Keep tracking of all nodes replaced in one execution
+	LAST_VERSION_REPLACED = (); // Keep tracking of the new version of a given node
+
+	ADJACENCIES_MATRIX = createAdjacenciesMatrix(flowGraph); // Mainly used to rebuild the renamed flow graph
+	IDOM_TREE = createIdomTree(createDominanceTree(flowGraph)); // Used to traverse the flow graph
+
+	map[Node, list[Node]] newBlockTree = replace(entryNode()); // Start algorithm
+
+	// Rebuild renamed flowGraph
+	FlowGraph newFlowGraph = {};
+	for(fatherNode <- newBlockTree) {
+		newFlowGraph = newFlowGraph + { <fatherNode, nodeChild> | nodeChild <- ADJACENCIES_MATRIX[fatherNode]};
+	};
+
+	return newFlowGraph;
+}
+
+public map[Node, list[Node]] replace(Node X) {
+	if((X == exitNode())) return ADJACENCIES_MATRIX;
+
+	Node oldNode = X;
+
+	// Deal with all nodes that aren't assigments bug uses a variable in some way
+	if(isNonAssignmentStatementToRename(X)) {
+		stmtNode(statement) = X;
+		Node renamedStatement = stmtNode(replaceImmediateUse(statement));
+		renameNodeOcurrecies(X, renamedStatement);
+		X = renamedStatement;
+	};
+
+	// Deal will all nodes that are an assigment and are not renamed
+	if(isOrdinaryAssignment(X) && !isRenamed(X)) {
+
+		// Replace right hand side variables
+		list[Immediate] rightHandNodeImmediates = getRightHandSideImmediates(X);
+		for(rightHandSideImmediate <- rightHandNodeImmediates) {
+			newAssignStmt = replaceRightVariableVersion(ADJACENCIES_MATRIX, rightHandSideImmediate, X);
+
+			renameNodeOcurrecies(X, newAssignStmt);
+
+			X = newAssignStmt;
+		};
+
+		// Replace left hand side variables
+		if(isLeftHandSideVariable(X)) {
+			Variable V = getStmtVariable(X);
+			Immediate localVariableImmediate = local(V[0]);
+
+			int assingmentQuantity = returnAssignmentQuantity(localVariableImmediate);
+			newAssignStmt = replaceLeftVariableVersion(ADJACENCIES_MATRIX, X, assingmentQuantity);
+
+			renameNodeOcurrecies(X, newAssignStmt);
+
+			stackVariableVersion(localVariableImmediate, assingmentQuantity);
+			iterateAssignmentQuantity(localVariableImmediate);
+			findAndAddPhiFunctionArgs(oldNode, newAssignStmt); // Search variable uses of replaced variable in phi-function and rename it
+
+			X = newAssignStmt;
+		};
+	}
+
+	for(child <- IDOM_TREE[X]) {
+		// We need to check the last version replaced because the renamed version was not beign reflected in the current `for` statement iteration,
+		// so we need to check the last version replaced
+		nodeToRename = LAST_VERSION_REPLACED[child]? ? LAST_VERSION_REPLACED[child] : child;
+		replace(nodeToRename);
+	};
+
+	if(!ignoreNode(oldNode) && isVariable(oldNode) && !isRenamed(X))
+		popOldNode(oldNode);
+
+	return ADJACENCIES_MATRIX;
+}
```

Comment:
This is the largest module and the most messy one, but try to mimic the Cytron's algorithm: 

![Screen Shot 2021-03-23 at 20 08 26](https://user-images.githubusercontent.com/14284674/112230518-b7bfce00-8c13-11eb-8137-5fd9c0274416.png)